PRAKTIKUM 12

In [2]:
# dataset passengers

#NAMA ICHA NURA NUGRAHA
#KELAS S1-INFORMATIKA BDL 5
#NIM 19.11.2646
import pymongo
import pandas as pd
import numpy as np
client = pymongo.MongoClient('mongodb://localhost:27017/')
db = client['titanic']
print(db.list_collection_names())



['passengers']


In [3]:
# Membuat dataframe dari collections 
passengers = db.passengers

res = passengers.find()
df_passengers = pd.DataFrame(list(res))

In [4]:
#  Mengecek apakah ada nilai NaN pada dataframe passengers
df_passengers.isna()

_id  PassengerId  Survived  Pclass   Name    Sex    Age  SibSp  Parch  \
0    False        False     False   False  False  False  False  False  False   
1    False        False     False   False  False  False  False  False  False   
2    False        False     False   False  False  False  False  False  False   
3    False        False     False   False  False  False  False  False  False   
4    False        False     False   False  False  False  False  False  False   
..     ...          ...       ...     ...    ...    ...    ...    ...    ...   
886  False        False     False   False  False  False  False  False  False   
887  False        False     False   False  False  False  False  False  False   
888  False        False     False   False  False  False   True  False  False   
889  False        False     False   False  False  False  False  False  False   
890  False        False     False   False  False  False  False  False  False   

     Ticket   Fare  Embarked  Cabin  
0     False  False     False   True  
1     False  False     False  False  
2     False  False     False   True  
3     False  False     False  False  
4     False  False     False   True  
..      ...    ...       ...    ...  
886   False  False     False   True  
887   False  False     False  False  
888   False  False     False   True  
889   False  False     False  False  
890   False  False     False   True  

[891 rows x 13 columns]

In [5]:
# Menghitung jumlah missing value dari setiap kolom
df_passengers.isna().sum()

_id              0
PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Embarked         2
Cabin          687
dtype: int64

In [6]:
# Melihat data type yang nanti akan kita gunakan
df_passengers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 13 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   _id          891 non-null    object
 1   PassengerId  891 non-null    object
 2   Survived     891 non-null    object
 3   Pclass       891 non-null    object
 4   Name         891 non-null    object
 5   Sex          891 non-null    object
 6   Age          714 non-null    object
 7   SibSp        891 non-null    object
 8   Parch        891 non-null    object
 9   Ticket       891 non-null    object
 10  Fare         891 non-null    object
 11  Embarked     889 non-null    object
 12  Cabin        204 non-null    object
dtypes: object(13)
memory usage: 90.6+ KB


In [7]:
#mengubah type data Age menjadi float
df_passengers['Age']=df_passengers['Age'].astype(float)

In [8]:
df_passengers.head()

_id PassengerId Survived Pclass  \
0  60d04ebfdb8077b4ebc9865f           1        0      3   
1  60d04ebfdb8077b4ebc98660           2        1      1   
2  60d04ebfdb8077b4ebc98661           3        1      3   
3  60d04ebfdb8077b4ebc98662           4        1      1   
4  60d04ebfdb8077b4ebc98663           5        0      3   

                                                Name     Sex   Age SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0     1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0     1   
2                             Heikkinen, Miss. Laina  female  26.0     0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0     1   
4                           Allen, Mr. William Henry    male  35.0     0   

  Parch            Ticket     Fare Embarked Cabin  
0     0         A/5 21171     7.25        S   NaN  
1     0          PC 17599  71.2833        C   C85  
2     0  STON/O2. 3101282    7.925        S   NaN  
3     0            113803     53.1        S  C123  
4     0            373450     8.05        S   NaN

In [9]:
# melakukan pengecheckan setelah melakukan perubahan 
df_passengers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 13 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   _id          891 non-null    object 
 1   PassengerId  891 non-null    object 
 2   Survived     891 non-null    object 
 3   Pclass       891 non-null    object 
 4   Name         891 non-null    object 
 5   Sex          891 non-null    object 
 6   Age          714 non-null    float64
 7   SibSp        891 non-null    object 
 8   Parch        891 non-null    object 
 9   Ticket       891 non-null    object 
 10  Fare         891 non-null    object 
 11  Embarked     889 non-null    object 
 12  Cabin        204 non-null    object 
dtypes: float64(1), object(12)
memory usage: 90.6+ KB


In [10]:
# identifikasi NaN values menggunakan boolean index
print(df_passengers[df_passengers['Age'].isnull()])

                          _id PassengerId Survived Pclass  \
5    60d04ebfdb8077b4ebc98664           6        0      3   
17   60d04ebfdb8077b4ebc98670          18        1      2   
19   60d04ebfdb8077b4ebc98672          20        1      3   
26   60d04ebfdb8077b4ebc98679          27        0      3   
28   60d04ebfdb8077b4ebc9867b          29        1      3   
..                        ...         ...      ...    ...   
859  60d04ebfdb8077b4ebc989ba         860        0      3   
863  60d04ebfdb8077b4ebc989be         864        0      3   
868  60d04ebfdb8077b4ebc989c3         869        0      3   
878  60d04ebfdb8077b4ebc989cd         879        0      3   
888  60d04ebfdb8077b4ebc989d7         889        0      3   

                                         Name     Sex  Age SibSp Parch  \
5                            Moran, Mr. James    male  NaN     0     0   
17               Williams, Mr. Charles Eugene    male  NaN     0     0   
19                    Masselmani, Mrs. Fatima

DARI IDENTIFIKASI KITA DAPAT MELIHAT KOLOM AGE KOSONG PADA BARIS 5 , 17 , 19 , 26 , 28 
... 859 , 863 , 868 , 878 , 888 

In [9]:
# MENCARI NILAI RATA RATA PADA KOLOM UMUR
rata_umur = df_passengers['Age'].mean()
print(rata_umur)

29.69911764705882


In [10]:
# MENGISI NILAI KOSONG PADA KOLOM AGE DENGAN NILAI RATA RATA YANG DIDAPAT
# MENGGUNAKAN FUNGSI FILLNA() DENGAN MENGGUNAKAN NILAI YANG KITA TENTUKAN
df_passengers['Age'] = df_passengers['Age'].fillna(rata_umur)

In [12]:
#KITA DATA KOSONG SUDAH BERISIKAN NILAI RATA RATA , DISETIAP BARIS YANG SUDAH DI IDENTIFIKASI
df_passengers.loc[df_passengers['Age']==29.69911764705882,['Age']]

Age
5    29.699118
17   29.699118
19   29.699118
26   29.699118
28   29.699118
..         ...
859  29.699118
863  29.699118
868  29.699118
878  29.699118
888  29.699118

[177 rows x 1 columns]

In [13]:
# MELAKUKAN PENGECHECKAN MISSING VALUE PADA KOLOM AGE
df_passengers['Age'].isna().sum()

0

In [14]:
# MELAKUKAN PENGECHECKAN MISSING VALUE 
# PADA DATAFRAME MASIH TERDAPAT MISSING VALUE PADA Embarked dan Cabin
df_passengers.isna().sum()

_id              0
PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age              0
SibSp            0
Parch            0
Ticket           0
Fare             0
Embarked         2
Cabin          687
dtype: int64

LALU KITA MENCOBA MELAKUKAN PERBANDINGAN PENGISIAN DATA KOSONG DENGAN NILAI TENGAH

In [11]:
# MELAKUKAN PENCARIAN NILAI TENGAH 
Umur_tengah = df_passengers['Age'].median()
print(Umur_tengah)

28.0


KITA MASUKAN NILAI TENGAH YANG SUDAH KITA CARI 

In [12]:
df_passengers['Age'] = df_passengers['Age'].fillna(Umur_tengah)

In [13]:
#KITA CHECK NILAI MASIH TERDAPAT NILAI KOSONG ATAU TIDAK
df_passengers.isna().sum()

_id              0
PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age              0
SibSp            0
Parch            0
Ticket           0
Fare             0
Embarked         2
Cabin          687
dtype: int64

In [14]:
#KITA CHECK DATA KOSONG SUDAH TER ISI DENGAN NILAI TENGAH YANG KITA CARI TADI
df_passengers.loc[df_passengers['Age']==28.0,['Age']]

Age
5    28.0
17   28.0
19   28.0
23   28.0
26   28.0
..    ...
868  28.0
874  28.0
878  28.0
883  28.0
888  28.0

[202 rows x 1 columns]

ANALISA PENGISIAN DATA KOSONG ANTARA RATA-RATA UMUR DAN NILAI TENGAH DAPAT DISIMPULKAN BAHWA RATA RATA MENGISI UMUR DENGAN NILAI 29.69911764705882 SEDANGKAN NILAI TENGAH 28 JADI RELATIF LEBIH BAIK MENGGUNAKAN MEDIAN ( NILAI TENGAH ) DALAM MENGISI MISSING VALUE PADA KOLOM AGE KARENA UMUR MERUPAKAN NUMERIK DISKRET 

In [15]:
# feature Cabin merupakan variabel kategorik, saya akan mengisi dengan nilai modus
# mengisi data kosong pada kolom cabin menggunakan nilai modus
import statistics as stat
df_passengers['Cabin'].fillna(stat.mode(df_passengers['Cabin'].dropna()), inplace = True)

In [16]:
df_passengers.isnull().sum()

_id            0
PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Embarked       2
Cabin          0
dtype: int64

In [17]:
print(df_passengers)

                          _id PassengerId Survived Pclass  \
0    60d04ebfdb8077b4ebc9865f           1        0      3   
1    60d04ebfdb8077b4ebc98660           2        1      1   
2    60d04ebfdb8077b4ebc98661           3        1      3   
3    60d04ebfdb8077b4ebc98662           4        1      1   
4    60d04ebfdb8077b4ebc98663           5        0      3   
..                        ...         ...      ...    ...   
886  60d04ebfdb8077b4ebc989d5         887        0      2   
887  60d04ebfdb8077b4ebc989d6         888        1      1   
888  60d04ebfdb8077b4ebc989d7         889        0      3   
889  60d04ebfdb8077b4ebc989d8         890        1      1   
890  60d04ebfdb8077b4ebc989d9         891        0      3   

                                                  Name     Sex   Age SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0     1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0     1   
2                               H

KESIMPULAN

MISSING VALUE MERUPAKAN SUATU INFORMASI YANG TIDAK TERSEDIA MEMILIKI NILAI KOSONG NaN PADA SUATU OBJEK DALAM SETIAP KASUS , MISSING VALUE TERJADI KARENA INFORMASI TERSEBUT TIDAK DIBERIKAN ATAU , SULIT UNTUK DICARI BAHKAN KESALAHAN SUATU SYSTEM YANG MENYEBABKAN TERJADINYA DATA INFORMASI TERSEBUT KOSONG . DALAM MELAKUKAN PENGISIAN DAN PENGHAPUSAN MISSING VALUE TIDAK SEMUA DATA HARUS DI ISI ADA BEBERAPA ALGORITMA YANG TIDAK PERLU UNTUK MELAKUKAN PREPORCESING DATA SEPERTI KKN DECISION TREE DLL YANG TIDAK MENGHARUSKAN PENGHAPUSAN DAN PENGISIAN VALUE DATA SEPERTI MELAKUKAN REGRESION LINEAR , PENGISIAN MISSING VALUE DAPAT DISESUAIKAN DENGAN KEBUTUHAN SEPERTI PENGISIAN CONTOH DIATAS DALAM MENANGANI KASUS MISSING VALUE 'AGE' DENGAN MENGGUNAKAN NILAI MEDIAN SEBAGAI VARIBLE NILAI DISKRET DAN PENGISIAN PADA CABIN YANG MENGGUNAKAN NILAI MODULUS SEBAGAI VARIBLE NILAI KATEGORIK